In [3]:
import pandas as pd
import numpy as np

In [11]:
url='https://raw.githubusercontent.com/Nikhil14384/Feature_Engineering/main/train_ohe.csv'

In [17]:
df=pd.read_csv(url,usecols=['X1','X2','X3','X4','X5','X6'])

In [18]:
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [ ]:
#lets check how namy labels in each column
for col in df:
  print(col,' : ',len(df[col].unique()),'labels')

lets see how many cols will get for the 6 cols we have if we perform ohe.



In [25]:
#ohe on 6 cols
pd.get_dummies(df,drop_first=True).shape

(4209, 117)

if we observe we got 117 cols just for 6 cols if we do ohe, just imagine for real life complex datasets. So we will be using a unique technique within ohe to get better results with very less columns.

In [28]:
#top 10 most frequent categories for X2 column
df.X2.value_counts().sort_values(ascending=False).head(20)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
aq      63
ay      54
a       47
t       29
k       25
i       25
b       21
ao      20
z       19
ag      19
Name: X2, dtype: int64

In [31]:
#make a list with the most frequent categories(10)
t10=[x for x in df.X2.value_counts().sort_values(ascending=False).head(10).index]
t10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [35]:
#we makethe entire dataframe into o's and 1's
for label in t10:
  df[label]=np.where(df['X2']==label,1,0)

In [ ]:
df[['X2']+t10].head(20)

Like this we can do for each and every column and ends up having 60 cols(10 for each). it will give the better results as well.